File: Learning_Japan_project_fe/learning-app-fe/src/services/assessmentItemService.ts
import { http } from "@/lib/http";
import { API_ENDPOINTS } from "@/config/api";
import { AssessmentType } from "@/enums/assessmentType";
export interface AssessmentItemResponse {
    id: string;
    name: string;
    level: string;
    questionCount: number;
    pointPerQuestion: number;
    totalPoint: number;
    assessmentType: AssessmentType;
    createdAt: string;
    updatedAt: string;
}

export interface UpdateAssessmentItemRequest {
    sectionId?: string;
    name?: string;
    level?: string;
    questionCount?: number;
    pointPerQuestion?: number;
    assessmentType?: AssessmentType;
}
export const assessmentItemService = {
    /* ================= GET BY SECTION ================= */
    getBySection: async (sectionId: string) => {
        return http.get<AssessmentItemResponse[]>(
            API_ENDPOINTS.ASSESSMENT_ITEM.BY_SECTION(sectionId)
        );
    },
    getAll: async () => {
        return http.get<AssessmentItemResponse[]>(
            API_ENDPOINTS.ASSESSMENT_ITEM.GET_ALL
        );
    },
    /* ================= GET DETAIL ================= */
    getDetail: async (id: string) => {
        return http.get<AssessmentItemResponse>(
            API_ENDPOINTS.ASSESSMENT_ITEM.DETAIL(id)
        );
    },

    /* ================= UPDATE ================= */
    update: async (
        id: string,
        data: UpdateAssessmentItemRequest
    ) => {
        return http.put(
            API_ENDPOINTS.ASSESSMENT_ITEM.UPDATE(id),
            data
        );
    },
};

File: Learning_Japan_project/learningApp/src/main/java/com/example/learningApp/service/assessment/AssessmentItemService.java
package com.example.learningApp.service.assessment;

import com.example.learningApp.common.EntityFinder;
import com.example.learningApp.dto.request.exam.assessment.UpdateAssessmentItemRequest;
import com.example.learningApp.dto.response.exam.assessment.AssessmentItemResponse;
import com.example.learningApp.entity.AssessmentItem;
import com.example.learningApp.entity.ExamSection;
import com.example.learningApp.mapper.AssessmentItemMapper;
import com.example.learningApp.repository.AssessmentItemRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
public class AssessmentItemService {

    private final AssessmentItemRepository repository;
    private final EntityFinder finder;
    @Qualifier("assessmentItemMapperImpl")
    private final AssessmentItemMapper mapper;

    @Transactional(readOnly = true)
    public List<AssessmentItemResponse> getAll() {
        return repository.findAll()
                .stream()
                .map(mapper::toResponse)
                .toList();
    }

    @Transactional(readOnly = true)
    public List<AssessmentItemResponse> getByLevel(String level) {
        return repository.findByLevel(level)
                .stream()
                .map(mapper::toResponse)
                .toList();
    }

    @Transactional(readOnly = true)
    public List<AssessmentItemResponse> getBySection(String sectionId) {
        ExamSection section = finder.examSectionId(sectionId);
        return repository.findBySection(section)
                .stream()
                .map(mapper::toResponse)
                .toList();
    }

    @Transactional(readOnly = true)
    public AssessmentItemResponse getDetail(String itemId) {
        AssessmentItem item = finder.assessmentItemId(itemId);
        return mapper.toResponse(item);
    }

    @Transactional
    public String update(String itemId, UpdateAssessmentItemRequest request) {
        AssessmentItem item = finder.assessmentItemId(itemId);

        if (request.getSectionId() != null) {
            ExamSection section = finder.examSectionId(request.getSectionId());
            item.setSection(section);
        }

        if (request.getName() != null)
            item.setName(request.getName());
        if (request.getLevel() != null)
            item.setLevel(request.getLevel());
        if (request.getQuestionCount() != null)
            item.setQuestionCount(request.getQuestionCount());
        if (request.getPointPerQuestion() != null)
            item.setPointPerQuestion(request.getPointPerQuestion());
        if (request.getAssessmentType() != null)
            item.setAssessmentType(request.getAssessmentType());

        if (item.getQuestionCount() != null && item.getPointPerQuestion() != null) {
            item.setTotalPoint(item.getQuestionCount() * item.getPointPerQuestion());
        }

        return "Update assessment item successfully";
    }

    @Transactional
    public String deleteByLevel(String level) {
        repository.deleteByLevel(level);
        return "Delete all assessment items of level " + level + " successfully";
    }
}